pkg/serviceaccounttoken/secret.go

```diff
@@ -117,13 +127,22 @@ func serviceAccountSecretPrefix(sa *v1.ServiceAccount) string {
 	return fmt.Sprintf("%s-token-", sa.Name)
 }
 
-// ServiceAccountSecretName returns the secret name for the given Service Account.
+// ServiceAccountSecret returns the secret for the given Service Account.
 // If there are more than one, it returns the first.
-func ServiceAccountSecretName(sa *v1.ServiceAccount) string {
-	if len(sa.Secrets) < 1 {
-		return ""
+func ServiceAccountSecret(sa *v1.ServiceAccount, secretLister secretLister) (*v1.Secret, error) {
+	if sa == nil {
+		return nil, fmt.Errorf("cannot get secret for nil service account")
+	}
+	secrets, err := secretLister(sa.Namespace, labels.SelectorFromSet(map[string]string{
+		ServiceAccountSecretLabel: sa.Name,
+	}))
+	if err != nil {
+		return nil, fmt.Errorf("could not get secrets for service account: %w", err)
+	}
+	if len(secrets) < 1 {
+		return nil, nil
 	}
-	return sa.Secrets[0].Name
+	return secrets[0], nil
```

Comment:
Question:
Is it possible that some secrets are left in the cluster and never deleted because we no longer store the secret reference on the service account and always return the 1st item? 

Or it will not happen because the names of the service accounts are defined ( i.e. not random values), `EnsureSecretForServiceAccount`  will eventually remove the "invalid" secret eventually after a few reconciliations?

Reply:
> Is it possible that some secrets are left in the cluster and never deleted

Yes, this would leave the existing secrets that don't have the label on them as they are. But, this function would never return those secrets, it would only find the new secrets that have the label.

> Or it will not happen because the names of the service accounts are defined

There is no guarantee what the service account will be named, this is just a utility package and it is up to the caller to name the service account.

> EnsureSecretForServiceAccount will eventually remove the "invalid" secret eventually after a few reconciliations?

`EnsureSecretForServiceAccount` most of the time will never delete secrets, it's not meant as a cleanup utility. I put the delete there to handle the unlikely case that a labeled secret is found for that SA that is not a service account token or does not have the SA annotation, which should only happen if someone manually changed or created it.

The warning only appears when the legacy token secret is being used to authenticate, so having the old secrets around should not make the warning appear unless we are using them. Do you think we need a migration to delete those old secrets?

Reply:
Hi @cmurphy, thank you for the explanation. 

AFAIK, in the case of provisioning v2, the secret is used in the system-agent install script during the bootstrapping progress, so my understanding is that the token is used every time a new node is added to the cluster. Therefore, I think we should have a migration logic. 

I would like to ping @Oats87 @jakefhyde for confirmation.

Reply:
@jiaqiluo Can you clarify what you mean by "migration logic"? Meaning, which of the following (if any) do you expect:
- Existing tokens should continue to work as-is, but the `serviceAccount.Secrets` should be updated so that use of these tokens doesn't produce warning (if so, when should this happen? On the next ensure? On start of Rancher?)
- Existing tokens should be removed as-soon-as-possible, the code is robust enough to request a new token, and we want to only use tokens that have been newly generated (so that they don't produce errors)

Reply:
hi @MbolotSuse, 
My initial thought is similar to the 2nd approach in your list. 
Whether it is doable depends on how the service account token is used and preserved during the lifecycle of a provisioning v1 cluster. Therefore, we need @Oats87 and @jakefhyde  for more input before we can make a decision.

Reply:
@cmurphy @MbolotSuse @jiaqiluo I personally tested this branch and can confirm it works for provisioning both fresh clusters and after upgrading to v1.27.x.